src/platform/video/ffmpegvideostream.cpp

```diff
@@ -0,0 +1,191 @@
+#include "videobuffer.h"
+
+extern "C" {
+#include <libavcodec/avcodec.h>
+#include <libswscale/swscale.h>
+}
+
+#include <captainslog.h>
+
+namespace Thyme
+{
+
+FFmpegVideoStream::FFmpegVideoStream(FFmpegFile *file) : m_ffmpegFile(file)
+{
+    file->Set_Frame_Callback(On_Frame);
+    file->Set_User_Data(this);
+    // Decode until we have our first video frame
+    while (m_good && m_frame == nullptr)
+        m_good = m_ffmpegFile->Decode_Packet();
+}
+
+FFmpegVideoStream::~FFmpegVideoStream()
+{
+    if (m_swsContext != nullptr) {
+        sws_freeContext(m_swsContext);
+        m_swsContext = nullptr;
+    }
+
+    if (m_ffmpegFile != nullptr) {
+        delete m_ffmpegFile;
+        m_ffmpegFile = nullptr;
+    }
+}
+
+void FFmpegVideoStream::On_Frame(AVFrame *frame, int stream_idx, int stream_type, void *user_data)
+{
+    FFmpegVideoStream *stream = static_cast<FFmpegVideoStream *>(user_data);
+    if (stream_type == AVMEDIA_TYPE_VIDEO) {
+        stream->m_frame = frame;
+    }
+}
+
+/**
+ * Update the stream state.
+ */
+void FFmpegVideoStream::Update()
+{
+    // BinkWait(m_binkHandle);
+}
+
+/**
+ * Check if a frame is ready to be drawn.
+ */
+bool FFmpegVideoStream::Is_Frame_Ready()
+{
+    return m_good && m_frame != nullptr;
+}
+
+/**
+ * Decompress a frame ready to render to a video buffer.
+ */
+void FFmpegVideoStream::Decompress_Frame()
+{
+    // Our frames are always decoded / decompressed. We might want to change at some point
+}
+
+/**
+ * Render a frame to a video buffer.
+ */
+void FFmpegVideoStream::Render_Frame(VideoBuffer *buffer)
+{
+    if (buffer == nullptr) {
+        return;
+    }
+
+    if (m_frame == nullptr) {
+        return;
+    }
+
+    if (m_frame->data == nullptr) {
+        return;
+    }
+
+    AVPixelFormat dst_pix_fmt;
+
+    switch (buffer->Get_Format()) {
+        case VideoBuffer::TYPE_R8G8B8:
+            dst_pix_fmt = AV_PIX_FMT_RGB24;
+            break;
+        case VideoBuffer::TYPE_X8R8G8B8:
+            dst_pix_fmt = AV_PIX_FMT_BGR0;
+            break;
+        case VideoBuffer::TYPE_R5G6B5:
+            dst_pix_fmt = AV_PIX_FMT_RGB565;
+            break;
+        case VideoBuffer::TYPE_X1R5G5B5:
+            dst_pix_fmt = AV_PIX_FMT_RGB555;
+            return;
+        default:
+            return;
+    }
+
+    m_swsContext = sws_getCachedContext(m_swsContext,
+        Width(),
+        Height(),
+        (enum AVPixelFormat)m_frame->format,
+        buffer->Get_Width(),
+        buffer->Get_Height(),
+        dst_pix_fmt,
+        SWS_BICUBIC,
+        NULL,
+        NULL,
+        NULL);
+
+    uint8_t *buffer_data = static_cast<uint8_t *>(buffer->Lock());
+    if (buffer_data == nullptr) {
+        captainslog_error("Failed to lock videobuffer");
+        return;
+    }
+
+    int dst_strides[] = { (int)buffer->Get_Pitch() };
+    uint8_t *dst_data[] = { buffer_data };
+    int result = sws_scale(m_swsContext, m_frame->data, m_frame->linesize, 0, Height(), dst_data, dst_strides);
```

Comment:
How does this know size of arrays?

Reply:
This depends on the pixelformat. E.g. YUV420P has multiple planes (Y and UV planes), while RGB32 everything is stored in a linear buffer. Since all pixelformats we convert to are linear, we're always have a single buffer. 
Formats are:
- AV_PIX_FMT_RGB24
- AV_PIX_FMT_BGR0
- AV_PIX_FMT_RGB565
- AV_PIX_FMT_RGB555